torch/sparse/_semi_structured_conversions.py

```diff
@@ -0,0 +1,344 @@
+        )
+
+    if meta_dtype == torch.int32:
+        magic0 = 4
+        magic1 = [0, 1, 32, 33]
+    elif meta_dtype == torch.int16:
+        magic0 = 8
+        magic1 = [0, 1, 4, 5]
+    tmp1 = torch.tensor([0, 2], dtype=torch.int64, device=device).repeat(
+        meta_nrows, meta_ncols // 2
+    )
+    tmp2 = (
+        (torch.arange(0, meta_ncols // 2, device=device) * 2 * meta_nrows)
+        .view(-1, 1)
+        .repeat(1, 2)
+        .view(-1)
+        .repeat(m, 1)
+    )
+    tmp3 = (
+        (torch.arange(0, 8, device=device) * magic0)
+        .view(-1, 1)
+        .repeat(m // 8, meta_ncols)
+    )
+    tmp4 = (
+        torch.tensor(magic1, device=device)
+        .view(-1, 1)
+        .repeat(1, 8 * meta_ncols)
+        .repeat(meta_nrows // 32, 1)
+        .view(meta_nrows, meta_ncols)
+    )
+    tmp5 = (
+        (torch.arange(0, meta_nrows // 32, device=device) * 64)
+        .view(-1, 1)
+        .repeat(1, 32 * meta_ncols)
+        .view(meta_nrows, meta_ncols)
+    )
+    meta_offsets = tmp1 + tmp2 + tmp3 + tmp4 + tmp5
+
+    meta = torch.gather(meta_reordered.view(-1), 0, meta_offsets.view(-1)).view(
+        m, meta_ncols
+    )
+
+    meta_2 = torch.empty(
+        (m, meta_ncols, 2 * quadbits_per_meta_elem), dtype=meta_dtype, device=device
+    )
+    if quadbits_per_meta_elem == 4:
+        meta_2[:, :, 0] = meta & 0b11
+        meta_2[:, :, 1] = (meta >> 2) & 0b11
+        meta_2[:, :, 2] = (meta >> 4) & 0b11
+        meta_2[:, :, 3] = (meta >> 6) & 0b11
+        meta_2[:, :, 4] = (meta >> 8) & 0b11
+        meta_2[:, :, 5] = (meta >> 10) & 0b11
+        meta_2[:, :, 6] = (meta >> 12) & 0b11
+        meta_2[:, :, 7] = (meta >> 14) & 0b11
+    elif quadbits_per_meta_elem == 8:
+        meta_2[:, :, 0] = meta & 0b11
+        meta_2[:, :, 1] = (meta >> 2) & 0b11
+        meta_2[:, :, 2] = (meta >> 4) & 0b11
+        meta_2[:, :, 3] = (meta >> 6) & 0b11
+        meta_2[:, :, 4] = (meta >> 8) & 0b11
+        meta_2[:, :, 5] = (meta >> 10) & 0b11
+        meta_2[:, :, 6] = (meta >> 12) & 0b11
+        meta_2[:, :, 7] = (meta >> 14) & 0b11
+        meta_2[:, :, 8] = (meta >> 16) & 0b11
+        meta_2[:, :, 9] = (meta >> 18) & 0b11
+        meta_2[:, :, 10] = (meta >> 20) & 0b11
+        meta_2[:, :, 11] = (meta >> 22) & 0b11
+        meta_2[:, :, 12] = (meta >> 24) & 0b11
+        meta_2[:, :, 13] = (meta >> 26) & 0b11
+        meta_2[:, :, 14] = (meta >> 28) & 0b11
+        meta_2[:, :, 15] = (meta >> 30) & 0b11
+
+    dense_offsets = meta_2.view(-1) + (
+        torch.arange(0, m * k // 2, device=device) * 4
+    ).view(-1, 1).repeat(1, 2).view(-1)
+
+    dense = torch.zeros((m * 2 * k,), dtype=sparse.dtype, device=device)
+    dense.scatter_(0, dense_offsets, sparse.view(-1))
+
+    return dense.view(m, 2 * k)
+
+
+def sparse_semi_structured_from_dense(dense):
+    aggressive_fusion_prev = cfg.aggressive_fusion
+    cfg.aggressive_fusion = True
```

Comment:
If you want to ensure that this config *only* applies to the function, then let's use torch.compile in a local context. And even then I'm not sure that'll perfectly work.

For now we need to rely on the user to set this. Obviously that doesn't scale (we can't just document that somewhere and expect people to read it).

However, we also can't ship this by turned on and compile'd by default. Let's do that in a separate PR and spawn some discussion with the torch.compile folks.

Reply:
Please remove this line and turn it into a comment.